groovy-language-server/src/main/java/com/palantir/ls/util/SourceWriter.java

```diff
@@ -0,0 +1,181 @@
+
+        // Check if any of the ranges intersect
+        for (int i = 0; i < sortedChanges.size() - 1; i++) {
+            Range range1 = sortedChanges.get(i).getRange();
+            Range range2 = sortedChanges.get(i + 1).getRange();
+            checkArgument(!Ranges.intersects(range1, range2),
+                    String.format("Cannot apply changes with intersecting ranges %s and %s", range1, range2));
+        }
+
+        String newFileContents = handleChanges(sortedChanges);
+
+        FileUtils.writeStringToFile(destination.toFile(), newFileContents, StandardCharsets.UTF_8.toString());
+    }
+
+    private String handleChanges(List<TextDocumentContentChangeEvent> sortedChanges) throws IOException {
+        BufferedReader file =
+                new BufferedReader(new InputStreamReader(new FileInputStream(destination.toAbsolutePath().toString()),
+                        StandardCharsets.UTF_8.toString()));
+        StringBuilder newFileContents = new StringBuilder();
+
+        int changeIdx = 0;
+        boolean endOfFile = false;
+        String line = file.readLine();
+        int lineNum = 0;
+        if (line == null) {
+            endOfFile = true;
+        }
+        int lastColumn = 0;
+
+        for (; !endOfFile && changeIdx < sortedChanges.size(); changeIdx++) {
+            Position start = sortedChanges.get(changeIdx).getRange().getStart();
+            Position end = sortedChanges.get(changeIdx).getRange().getEnd();
+
+            // Find the line where this change starts.
+            while (start.getLine() != lineNum) {
+                // Append this line which is not affected by a range.
+                newFileContents.append(line.substring(Math.min(lastColumn, line.length()))).append("\n");
+                lastColumn = 0;
+                line = file.readLine();
+                ++lineNum;
+                if (line == null) {
+                    endOfFile = true;
+                    break;
+                }
+            }
+
+            // Handle this change
+            if (endOfFile) {
+                break;
+            }
+            // Add everything before this range starting at the last column.
+            newFileContents.append(line.substring(Math.min(lastColumn, line.length()),
+                    Math.min(start.getCharacter(), line.length())));
+            lastColumn = 0;
+            newFileContents.append(sortedChanges.get(changeIdx).getText());
+
+            // Advance the file line buffer to where the range ends, ignoring all those lines.
+            while (end.getLine() != lineNum) {
+                line = file.readLine();
+                ++lineNum;
+                if (line == null) {
+                    endOfFile = true;
+                    break;
+                }
+            }
+
+            // Set the where the next line should start, which depends on if its an insertion or not.
+            lastColumn =
```

Comment:
Let's assume the end is always exclusive and make this
`lastColumn = end.getCharacter()`